compiler/kernel/expr.go

```diff
@@ -342,16 +343,20 @@ func (b *Builder) compileCall(call dag.Call) (expr.Evaluator, error) {
 	return expr.NewCall(fn, exprs), nil
 }
 
-func (b *Builder) compileUDFCall(name string, body dag.Expr) (expr.Function, error) {
+func (b *Builder) compileUDFCall(name string, f *dag.Func) (expr.Function, error) {
 	if fn, ok := b.compiledUDFs[name]; ok {
 		return fn, nil
 	}
-	fn := &expr.UDF{Name: name, Sctx: b.sctx()}
+	if b.udfStackSizeCounter == nil {
+		b.udfStackSizeCounter = new(atomic.Int64)
+		defer func() { b.udfStackSizeCounter = nil }()
+	}
+	fn := expr.NewUDF(b.sctx(), name, f.Params, b.udfStackSizeCounter)
```

Comment:
Sharing this counter across goroutines is surprising and doesn't feel like the right move.

Reply:
Open to suggestions. We are trying to get rid of using expr.Context which will give you use a solution that will also work in vector runtime. Since we are in expression land this shouldn't normally go across go-routines with the exception in some weird cases of unnest expressions and probably some less weird cases when we add subquery expressions (am I'm missing something here?). I don't think our current approach works in those circumstances either because once you go out of expression land into operator land and back the frame expr.Context is carried across these boundaries.

Reply:
> Open to suggestions. We are trying to get rid of using expr.Context which will give you use a solution that will also work in vector runtime. Since we are in expression land this shouldn't normally go across go-routines with the exception in some weird cases of unnest expressions and probably some less weird cases when we add subquery expressions (am I'm missing something here?).

I now have a better understanding of what's going on here and think it's fine because an expr.Evaluator should actually never be shared across goroutines. And given that, there's no need for a *atomic.Int64. A *int suffices.